go/tests/googleai_test.go

```diff
@@ -0,0 +1,165 @@
+		Models:        []string{"gemini-1.5-pro"},
+		Embedders:     []string{"embedding-001"},
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	t.Run("embed", func(t *testing.T) {
+		test, err := xltest.ReadFile[string, embedderResult](filepath.Join("testdata", "googleai-embedder.yaml"))
+		if err != nil {
+			t.Fatal(err)
+		}
+		testfunc := func(text string) (embedderResult, error) {
+			ts := &testTraceStore{}
+			remove := core.SetDevTraceStore(ts)
+			defer remove()
+			vals, err := ai.Embed(ctx, googleai.Embedder("embedding-001"), &ai.EmbedRequest{Document: ai.DocumentFromText(text, nil)})
+			if err != nil {
+				return embedderResult{}, err
+			}
+			return embedderResult{Values: vals, trace: ts.td}, nil
+		}
+		valfunc := validateEmbedder
+		if *update {
+			valfunc = updateEmbedder
+		}
+		test.Run(t, testfunc, valfunc, nil)
+	})
+}
+
+// validateEmbedder compares the results from running an embedder with
+// the desired results. The latter are taken from a YAML file and so are in raw
+// unmarshalled form.
+func validateEmbedder(got, want embedderResult) error {
+	if err := internal.ReadJSONFile(filepath.Join("testdata", want.TraceFile), &want.trace); err != nil {
+		return err
+	}
+	if !slices.Equal(got.Values, want.Values) {
+		return fmt.Errorf("values: got %v, want %v", got.Values, want.Values)
+	}
+	renameSpans(got.trace)
+	renameSpans(want.trace)
+	opts := []gocmp.Option{
+		cmpopts.IgnoreFields(tracing.Data{}, "TraceID", "StartTime", "EndTime"),
+		cmpopts.IgnoreFields(tracing.SpanData{}, "TraceID", "StartTime", "EndTime"),
+	}
+	if diff := gocmp.Diff(want.trace, got.trace, opts...); diff != "" {
+		return fmt.Errorf("traces: %s", diff)
+	}
+	return nil
+}
+
+// renameSpans changes the keys of td.Spans to s000, s001, ... in order of the span start time,
+// as well as references to those IDs within the spans.
+// This makes it possible to compare two span maps with different span IDs.
+func renameSpans(td *tracing.Data) {
+	type item struct {
+		id string
+		t  tracing.Milliseconds
+	}
+
+	var items []item
+	startTimes := map[tracing.Milliseconds]bool{}
+	for id, span := range td.Spans {
+		if startTimes[span.StartTime] {
+			panic("duplicate start times")
+		}
+		startTimes[span.StartTime] = true
+		items = append(items, item{id, span.StartTime})
+	}
+	slices.SortFunc(items, func(i1, i2 item) int {
+		return cmp.Compare(i1.t, i2.t)
+	})
+	oldIDToNew := map[string]string{}
+	for i, item := range items {
+		oldIDToNew[item.id] = fmt.Sprintf("s%03d", i)
+	}
+	// Re-create the span map with the new span IDs.
+	m := map[string]*tracing.SpanData{}
+	for oldID, span := range td.Spans {
+		newID := oldIDToNew[oldID]
+		if newID == "" {
+			panic(fmt.Sprintf("missing id: %q", oldID))
+		}
+		m[newID] = span
+		// A span references it own span ID and possibly its parent's.
+		span.SpanID = oldIDToNew[span.SpanID]
+		if span.ParentSpanID != "" {
+			span.ParentSpanID = oldIDToNew[span.ParentSpanID]
+		}
+	}
+	td.Spans = m
+}
```

Comment:
@pavelgj This code makes it possible to compare two traces completely, except for times and the trace ID.